BAU: Upgrade nimbusds

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Upgrade nimbusds

### Why did it change

This was a suggestion from dependabot, but there were a few breaking
changes to deal with.

The most notable is the removal of the inbuilt JSON lib that we were
using in places. This change replaces it with Jackson. It also allows us
to get rid of gson in a few places.

There are a few other small changes, such as the JWEDecrypter#decrypter
method having an extended signature, but nothing very big.

dependabot[bot] committed:
BAU: Bump com.nimbusds:oauth2-oidc-sdk from 9.27 to 11.10.1

Bumps [com.nimbusds:oauth2-oidc-sdk](https://bitbucket.org/connect2id/oauth-2.0-sdk-with-openid-connect-extensions) from 9.27 to 11.10.1.
- [Changelog](https://bitbucket.org/connect2id/oauth-2.0-sdk-with-openid-connect-extensions/src/master/CHANGELOG.txt)
- [Commits](https://bitbucket.org/connect2id/oauth-2.0-sdk-with-openid-connect-extensions/branches/compare/11.10.1..9.27)

---
updated-dependencies:
- dependency-name: com.nimbusds:oauth2-oidc-sdk
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Wynndow committed:
BAU: Upgrade nimbusds to 11.10.1

This was a suggestion from dependabot, but there were a few breaking
changes to deal with.

The most notable is the removal of the inbuilt JSON lib that we were
using in places. This change replaces it with Jackson. It also allows us
to get rid of gson in a few places.

There are a few other small changes, such as the JWEDecrypter#decrypter
method having an extended signature, but nothing very big.

Wynndow commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/service/KmsRsaDecrypter.java:
```diff
@@ -48,7 +48,8 @@ public byte[] decrypt(
             Base64URL cipherText,
-            Base64URL authTag)
+            Base64URL authTag,
+            byte[] aad)
```
This was a breaking change - the `aad` is a new parameter we need to include. We're not using it though so just need to update the signature to ensure we're implementing the interface.

Wynndow commented on lambdas/initialise-ipv-session/src/test/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandlerTest.java:
```diff
@@ -591,12 +591,14 @@ void shouldSendAuditEventForIpvInheritedIdentityVcReceived() throws Exception {
-                    "[{\"nameParts\":[{\"type\":\"GivenName\",\"value\":\"KENNETH\"},{\"type\":\"FamilyName\",\"value\":\"DECERQUEIRA\"}]}]",
-                    restricted.name().toString());
-            assertEquals("[{\"value\":\"1965-07-08\"}]", restricted.birthDate().toString());
+                    "[{\"nameParts\":[{\"value\":\"KENNETH\",\"type\":\"GivenName\"},{\"value\":\"DECERQUEIRA\",\"type\":\"FamilyName\"}]}]",
```
As we're now using Jackson, the ordering we set for fields in the `NameParts` class is now being respected, so this is just an ordering change.

Joe-Edwards-GDS replied:
Do you think it would be preferable to swap the order in the class instead? I think type, value is a more traditional order!

Wynndow replied:
I did wonder about this as I seem to remember it was me that set that ordering ages ago, for some reason. The examples in the vocab would seem to suggest value, then type: https://vocab.account.gov.uk/v1/examples/NameClass/OneGivenOneFamily.json
I'm not fussed really, happy to switch it back again if you think

Joe-Edwards-GDS replied:
:+1: matching the Vocab seems like as good a directive as any!

Wynndow commented on lambdas/process-cri-callback/src/main/java/uk/gov/di/ipv/core/processcricallback/service/CriApiService.java:
```diff
@@ -205,7 +205,7 @@ public VerifiableCredentialResponse fetchVerifiableCredential(
                         VerifiableCredentialResponse.builder()
                                 .verifiableCredentials(
-                                        Collections.singletonList(response.getContent()))
+                                        Collections.singletonList(response.getBody()))
```
`getContent()` has been deprecated in favour of `getBody()`

Wynndow commented on lambdas/process-cri-callback/src/main/java/uk/gov/di/ipv/core/processcricallback/service/CriCheckingService.java:
```diff
@@ -231,8 +231,10 @@ public JourneyResponse checkVcResponse(
 
-        if (!vcs.stream().allMatch(VcHelper::isSuccessfulVc)) {
-            return JOURNEY_FAIL_WITH_NO_CI;
+        for (var vc : vcs) {
```
`isSuccessfulVc` now throws a checked exception, so we can't easily use it in a stream

Joe-Edwards-GDS replied:
Javaaaaa :(

Wynndow replied:
![](https://media.giphy.com/media/l41lFU69ZXu03qWrK/giphy.gif?cid=790b7611lko3cxs40a0h1vfoyrd1laefktdpzavjwbz90ran&ep=v1_gifs_search&rid=giphy.gif&ct=g)

Wynndow commented on libs/audit-service/src/test/java/uk/gov/di/ipv/core/library/auditing/helpers/AuditExtensionsHelperTest.java:
```diff
@@ -50,23 +50,23 @@ void shouldGetVerifiableCredentialExtensionsForAudit() throws Exception {
         var restrictedData = getRestrictedAuditDataForF2F(PASSPORT_NON_DCMAW_SUCCESSFUL_VC);
         assertEquals(
-                "{\"name\":[{\"nameParts\":[{\"type\":\"GivenName\",\"value\":\"KENNETH\"},{\"type\":\"FamilyName\",\"value\":\"DECERQUEIRA\"}]}],\"docExpiryDate\":\"2030-01-01\"}",
+                "{\"name\":[{\"nameParts\":[{\"value\":\"KENNETH\",\"type\":\"GivenName\"},{\"value\":\"DECERQUEIRA\",\"type\":\"FamilyName\"}]}],\"docExpiryDate\":\"2030-01-01\"}",
```
Again, just reordering of the keys

Wynndow commented on libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/UserIdentityService.java:
```diff
@@ -191,10 +190,12 @@ public boolean isVcSuccessful(List<VcStatusDto> currentVcStatuses, String criIss
-                VcHelper.filterVCBasedOnProfileType(vcs, ProfileType.GPG45).stream()
-                        .filter(VcHelper::isSuccessfulVc)
-                        .toList();
+        var successfulVcs = new ArrayList<VerifiableCredential>();
```
Another case of not being able to use streams due do `isSuccessfulVc` throwing a checked exception.

Joe-Edwards-GDS commented on libs/verifiable-credentials/src/main/java/uk/gov/di/ipv/core/library/verifiablecredential/helpers/VcHelper.java:
```diff
@@ -211,13 +212,10 @@ private static Integer getAge(String dobValue) {
-        }
-        return docField;
+        var passport = credentialSubject.path(VC_PASSPORT);
+        return passport.isMissingNode() ? credentialSubject.path(VC_RESIDENCE_PERMIT) : passport;
```
You might now be able to simplify this to `credentialSubject.hasNonNull(VC_PASSPORT) ? credentialSubject.path(VC_PASSPORT) : credentialSubject.path(VC_RESIDENCE_PERMIT)`

(At which point you could consider inlining)

Wynndow replied:
ooo nice

Joe-Edwards-GDS commented on libs/verifiable-credentials/src/main/java/uk/gov/di/ipv/core/library/verifiablecredential/helpers/VcHelper.java:
```diff
@@ -75,12 +80,12 @@ public static boolean isSuccessfulVc(VerifiableCredential vc) {
+                    OBJECT_MAPPER.treeToValue(evidenceArray, CREDENTIAL_EVIDENCE_ITEM_LIST_TYPE));
+        } catch (JsonProcessingException e) {
+            throw new CredentialParseException("Unable to create credential evidence list", e);
+        }
```
It's a bit sad that this now throws an exception, but probably the best thing for now - we'll hopefully refactor when we introduce the vocab classes and rethink 'success' entirely...

Joe-Edwards-GDS commented on libs/audit-service/src/main/java/uk/gov/di/ipv/core/library/auditing/helpers/AuditExtensionsHelper.java:
```diff
@@ -24,64 +33,100 @@
         return new AuditExtensionsVcEvidence(
                 jwtClaimsSet.getIssuer(),
-                evidence,
+                evidenceArray.toString(),
```
This gets re-parsed in the `AuditExtensionsVcEvidence` constructor - can we change the type and avoid a round trip?

Wynndow replied:
oh yeah! done

Joe-Edwards-GDS reviewed (commented):
A few comments, but looks plausible to me

Wynndow committed:
BAU: Change AuditExtensionsVcEvidence constructor type

This sets the type of the evidence param to a JsonNode, which matches
the field. This means we don't have to do any convesrion. Which means we
don't throw an AuditExtensionException. Which means we can remove it, as
this was the only place it was used.

Joe-Edwards-GDS commented on libs/audit-service/src/main/java/uk/gov/di/ipv/core/library/auditing/extension/AuditExtensionsVcEvidence.java:
```diff
@@ -44,32 +42,21 @@ public class AuditExtensionsVcEvidence implements AuditExtensions {
 
-    public AuditExtensionsVcEvidence(String iss, String evidence, Boolean successful)
-            throws AuditExtensionException {
+    public AuditExtensionsVcEvidence(String iss, JsonNode evidence, Boolean successful) {
```
(I think this version is only used in a test and can probably be deleted as well, if you like!)

Joe-Edwards-GDS commented on libs/audit-service/src/main/java/uk/gov/di/ipv/core/library/auditing/extension/AuditExtensionsVcEvidence.java:
```diff
@@ -44,32 +42,21 @@ public class AuditExtensionsVcEvidence implements AuditExtensions {
+            @JsonProperty(value = "successful") Boolean successful,
+            @JsonProperty(value = "vot") Vot vot,
+            @JsonProperty(value = "isUkIssued") Boolean isUkIssued,
+            @JsonProperty(value = "age") Integer age) {
```
I'm a bit sceptical that we need these annotations, but happy to leave if there's any uncertainty.

(TBH this whole thing can probably just be a record class...)

Wynndow replied:
Yeah, i intentionally left them, mostly out of fear

Wynndow replied:
You've twisted my arm and I've gone the whole hog
